<?xml version="1.0" ?>
<launch>
  <arg name="rviz" default="false"/>

  <arg name="radius" default="0.25" />
  <arg name="height" default="0.1" />
  <arg name="max_range" default="4.5" />

  <arg name="resolution" default="0.1" />
  <arg name="depth_levels" default="16" />

  <!-- Depth parameters -->
  <arg name="insert_depth" default="1" />
  <arg name="clearing_depth" default="$(arg insert_depth)" />
  <arg name="publish_depth" default="4" />
  <arg name="collision_depth" default="$(arg insert_depth)" />
  <arg name="gain_depth" default="$(eval arg('collision_depth') + 1)" />

  <node pkg="nodelet" type="nodelet" name="ufoexplorer_nodelet_manager" args="manager" output="screen">
    <param name="num_worker_threads" value="12" />
  </node>

  <!-- Mapping -->
  <node pkg="ufomap_mapping" type="ufomap_mapping_server_node" name="ufomap_mapping_server_node" output="screen" required="true">
    <remap from="/cloud_in" to="/camera/depth/points2"/>

    <param name="multithreaded" value="true" />

    <param name="resolution" value="$(arg resolution)" />
    <param name="depth_levels" value="$(arg depth_levels)" />
    <param name="color_map" value="true"/>

    <param name="insert_depth" value="$(arg insert_depth)"/>
    <param name="simple_ray_casting" value="true"/>
    <param name="max_range" value="$(arg max_range)"/>

    <param name="prob_hit" value="0.8"/>
    <param name="prob_miss" value="0.45"/>

    <param name="clear_robot" value="false"/>
    <param name="robot_radius" value="$(eval arg('radius') + 0.15)"/>
    <param name="robot_height" value="$(eval arg('height') + 0.15)"/>
    <param name="clearing_depth" value="$(arg clearing_depth)"/>

    <param name="compress" value="false"/>
    <param name="update_part_of_map" value="true"/>
    <param name="update_rate" value="10.2"/>
    <param name="publish_depth" value="$(arg publish_depth)"/>
    <param name="pub_rate" value="0.0"/>
    <param name="map_queue_size" value="100" />
  </node>

  <node pkg="collision_avoidance" type="collision_avoidance_node" name="collision_avoidance_node" output="screen" required="true">
    <remap from="map" to="/ufomap_mapping_server_node/map_depth_$(arg collision_depth)" />
    <remap from="cloud" to="/camera/depth/points" />
    <remap from="odometry" to="/mavros/local_position/odom" />
    <remap from="~/control" to="/mavros/setpoint_velocity/cmd_vel" />
    <remap from="~/path" to="~/path" />
    <remap from="~/obstacles" to="~/obstacles" />
    <remap from="move_to" to="move_to" />

    <param name="num_cloud_topics" value="1" />

    <param name="robot_frame_id" value="base_stabilized_link" />

    <!-- Map -->
    <param name="map_frame_id" value="map"/>
    <param name="automatic_pruning" value="true"/>
    <param name="occupied_thres" value="0.5"/>
    <param name="free_thres" value="0.5"/>
    <param name="unknown_as_occupied" value="true"/>
    <param name="map_depth" value="$(arg collision_depth)"/>

    <param name="distance_converged" value="0.4" />
    <param name="yaw_converged" value="0.1" />

    <param name="frequency" value="20.0" />

    <param name="max_xy_vel" value="2.5" />
    <param name="max_z_vel" value="2.5" />
    <param name="max_yaw_rate" value="52.0" />
    <param name="h_m" value="1.0" />

    <param name="radius" value="$(arg radius)" />
    <param name="height" value="$(arg height)" />
    <param name="min_distance_hold" value="1.0" />

    <param name="security_distance" value="1.0" />
    <param name="epsilon" value="0.1" />

    <param name="polar_size" value="360" />

    <param name="max_times_backwards" value="10" />
    <param name="max_times_backwards_total" value="1000" />
    <param name="max_direction_change" value="60" />

    <param name="look_ahead_distance" value="2.0" />
    <param name="look_forward" value="false" />
    <param name="move_while_yawing" value="true" />
  </node>

  <node pkg="nodelet" type="nodelet" name="image_to_points" args="load depth_image_proc/point_cloud_xyzrgb ufoexplorer_nodelet_manager" required="true">
    <remap from="rgb/camera_info" to="/camera/depth/camera_info" />
    <remap from="rgb/image_rect_color" to="/camera/rgb/image_raw" />
    <remap from="depth_registered/image_rect" to="/camera/depth/image_raw2" />
    <remap from="depth_registered/points" to="/camera/depth/points2" />

    <param name="queue_size" value="100" />
  </node>

  <node pkg="nodelet" type="nodelet" name="image_maxing" args="load image_maxing/ImageMaxingNodelet ufoexplorer_nodelet_manager" output="screen" required="true">
    <remap from="input" to="/camera/depth/image_raw" />
    <remap from="output" to="/camera/depth/image_raw2" />

    <param name="value" value="$(eval arg('max_range') + arg('resolution'))" />
  </node>

  <!-- RViz -->
  <node pkg="rviz" type="rviz" name="rviz" if="$(arg rviz)" required="false" args="-d $(find exploration_planner)/rviz/view.rviz"/>
</launch>